# 3. Shanghai Rankings Exploratory Analysis

## Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## Importing dataset

In [2]:
# Set path variable

path = r'C:\Users\HP\Documents\CareerFoundry\Data Immersion\Achievement 6\World University Rankings Analysis'

In [3]:
# Import dataset

df_shanghai = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'shanghaiData.csv'))

In [4]:
# Check to see if it imported correctly

df_shanghai

world_rank                              university_name national_rank  \
0             1                           Harvard University             1   
1             2                      University of Cambridge             1   
2             3                          Stanford University             2   
3             4           University of California, Berkeley             3   
4             5  Massachusetts Institute of Technology (MIT)             4   
...         ...                                          ...           ...   
4892    401-500                        University of Trieste         11-20   
4893    401-500                       University of Zaragoza          9-13   
4894    401-500                        Utah State University       126-146   
4895    401-500              Vienna University of Technology           4-6   
4896    401-500                       Wake Forest University       126-146   

      total_score  alumni  award   hici     ns    pub   pcp  year  
0           100.0   100.0  100.0  100.0  100.0  100.0  72.4  2005  
1            73.6    99.8   93.4   53.3   56.6   70.9  66.9  2005  
2            73.4    41.1   72.2   88.5   70.9   72.3  65.0  2005  
3            72.8    71.8   76.0   69.4   73.9   72.2  52.7  2005  
4            70.1    74.0   80.6   66.7   65.8   64.3  53.0  2005  
...           ...     ...    ...    ...    ...    ...   ...   ...  
4892          NaN     0.0    0.0    5.0   10.9   25.1  20.1  2015  
4893          NaN     0.0    0.0    7.6    5.1   33.3  13.1  2015  
4894          NaN    13.6    0.0    3.6   10.8   25.1  15.5  2015  
4895          NaN     0.0    0.0    0.0   12.2   28.8  22.9  2015  
4896          NaN     0.0    0.0   14.9    7.5   25.0  11.9  2015  

[4897 rows x 11 columns]

In [5]:
# Check dimensions of dataframe

df_shanghai.shape

(4897, 11)

## Data Cleaning

### Dropping Irrelevant Columns

In [6]:
# Check the column names

df_shanghai.columns

Index(['world_rank', 'university_name', 'national_rank', 'total_score',
       'alumni', 'award', 'hici', 'ns', 'pub', 'pcp', 'year'],
      dtype='object')

It looks like most columns would be relevant for analysis. The rankings look to be based on certain criteria: alumni, award, hici, ns, and pub. 

### Renaming Columns

In [7]:
# Rename certain columns to be more intuitive

df_shanghai.rename(columns = {'world_rank' : 'world_rank_Shanghai',
                              'total_score' : 'overall_score_Shanghai',
                              'alumni' : 'alumni_score',
                              'award' : 'award_score',
                              'hici' : 'hici_score',
                              'ns' : 'ns_score',
                              'pub' : 'publication_score',
                              'pcp' : 'pcp_score'}, inplace = True)

In [8]:
# Check column names to see if they were replaced correctly

df_shanghai.columns

Index(['world_rank_Shanghai', 'university_name', 'national_rank',
       'overall_score_Shanghai', 'alumni_score', 'award_score', 'hici_score',
       'ns_score', 'publication_score', 'pcp_score', 'year'],
      dtype='object')

### Changing a Variable Data Type

In [9]:
# Check dtype of every column 

df_shanghai.dtypes

world_rank_Shanghai        object
university_name            object
national_rank              object
overall_score_Shanghai    float64
alumni_score              float64
award_score               float64
hici_score                float64
ns_score                  float64
publication_score         float64
pcp_score                 float64
year                        int64
dtype: object

Most of the columns seem to have appropriate datatypes aside from the 'world_rank_Shanghai' and 'national_rank' columns. It's expected that these columns should be integers.

In [10]:
# Set display option to show all rows. This will be useful for value_counts later 

pd.set_option('display.max_rows', None)

In [11]:
# Check value counts of 'world_rank_shanghai' column

df_shanghai['world_rank_Shanghai'].value_counts(dropna = False)

401-500    600
301-400    600
201-300    584
151-200    300
201-302    204
101-150    200
303-401    198
102-150    147
403-510    108
203-304    102
402-503    102
101-151    102
402-501    100
305-402     98
203-300     98
152-200     98
151-202     52
101-152     52
153-202     50
93          19
65          17
90          17
97          16
77          16
73          16
32          15
53          14
86          14
61          14
41          14
83          13
85          13
54          13
82          13
47          13
39          13
46          13
67          13
99          13
18          13
27          13
38          12
92          12
8           12
96          12
70          12
57          12
78          12
30          12
21          12
42          12
24          12
35          12
87          12
69          11
72          11
2           11
58          11
1           11
64          11
45          11
20          11
23          11
26          11
17          11
16          11
15        

While there are no blanks in this column, this column is unexpectedly under a "object" datatype because of the range for certain rows (ex: 401-500). It is impossible to know what the real ranking numbers are so I cannot simply replace them. It's fine to leave this column for now. If necessary, I can drop the rows that have these "ranges" so only schools with integer rankings remain. 

In [12]:
# Check value counts of 'national_rank' column

df_shanghai['national_rank'].value_counts(dropna = False)

1          343
2          206
3          133
4          122
1-2         86
2-3         84
5           69
5-7         65
5-6         64
7           58
6           51
4-5         50
126-146     42
3-5         42
71-90       40
1-4         40
24-33       40
3-4         38
2-4         38
4-6         36
7-9         36
6-7         36
8           35
5-9         34
9-13        34
55-70       32
20-30       31
10          31
88-118      31
55-69       30
9-18        30
91-119      29
89-117      29
15-24       29
141-168     28
23-36       28
15-23       27
6-8         27
78-104      27
141-167     27
14-22       27
111-137     27
113-138     26
112-137     26
141-166     26
115-139     25
79-102      24
13-18       24
91-114      24
7-18        24
8-9         24
1-3         24
9-14        24
5-8         24
8-13        24
118-140     23
103-125     23
91-112      22
90-111      22
23-33       22
119-140     22
90-110      21
120-140     21
69-89       21
8-14        21
105-125     21
11        

Similarly to the 'world_rank_Shanghai' column, this column is also filled with ranges which caused the entire column to be classified as an 'object' datatype. Another thing to note is that there is also one NaN value in this column. Aside from the ranges, the column is also a mix of floats and integers. I'll leave this column for now and may drop it later since national_rank may not lead to any proper insights. 

### Addressing Mixed Data Columns

In [13]:
# Check for mixed data types

for col in df_shanghai.columns.tolist():
  weird = (df_shanghai[[col]].applymap(type) != df_shanghai[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_shanghai[weird]) > 0:
    print (col)

university_name
national_rank


As mentioned above, these two columns have mixed datatypes because of the ranges and cannot be changed. For further analysis, it may be useful to have a subset of the data that does not include any of these ranges. 

### Addressing Missing Values

In [14]:
#Check for number of missing values in every column 

df_shanghai.isnull().sum()

world_rank_Shanghai          0
university_name              1
national_rank                1
overall_score_Shanghai    3796
alumni_score                 1
award_score                  2
hici_score                   2
ns_score                    22
publication_score            2
pcp_score                    2
year                         0
dtype: int64

There are missing values in the columns. While most of them can be ignored, it is important to note that the "overall_score_Shanghai" column has 3796 missing values in 4897 rows. This is over 75% of the values so this column may need to be dropped later on or a subset of this dataframe may need to be created. After looking at the universities that have these missing values, all of them had a rank that was below the top 100 (except for 1 case) so the lack of data may have been due to the rankings deeming universities outside of the top 100 unnecessary to have an overall score. 

In the case of the missing overall score that also had a world rank within the top 100, it comes from a university that had a rank of 99, but is missing all information except for the year and world rank. As a result, this row should be dropped. 

In [15]:
# Locate the row that has majority missing values

df_shanghai.loc[df_shanghai['university_name'].isnull()]

world_rank_Shanghai university_name national_rank  \
3896                  99             NaN           NaN   

      overall_score_Shanghai  alumni_score  award_score  hici_score  ns_score  \
3896                     NaN           NaN          NaN         NaN       NaN   

      publication_score  pcp_score  year  
3896                NaN        NaN  2013

In [16]:
# Drop the specified row

df_shanghai.dropna(subset = ['university_name'], inplace = True)

In [17]:
# Check the dimensions to see if the row was properly dropped
# The remaining dataframe should have one less row (4896 in total)

df_shanghai.shape

(4896, 11)

In [19]:
#Check remaining missing values 

df_shanghai.isnull().sum()

world_rank_Shanghai          0
university_name              0
national_rank                0
overall_score_Shanghai    3795
alumni_score                 0
award_score                  1
hici_score                   1
ns_score                    21
publication_score            1
pcp_score                    1
year                         0
dtype: int64

In [20]:
# Locate the row with missing value for award_score

df_shanghai.loc[df_shanghai['award_score'].isnull()]

world_rank_Shanghai       university_name national_rank  \
3797             201-300  University of Oregon        86-109   

      overall_score_Shanghai  alumni_score  award_score  hici_score  ns_score  \
3797                     NaN           9.0          NaN         NaN       NaN   

      publication_score  pcp_score  year  
3797                NaN        NaN  2012

This row also appears to have multiple missing values. In addition, it does not even break the top 100 universities so it should be dropped. 

In [21]:
# Drop the row and check to see if the shape of the dataframe changed

df_shanghai.dropna(subset = ['award_score'], inplace = True)

df_shanghai.shape

(4895, 11)

In [22]:
#Check remaining missing values

df_shanghai.isnull().sum()

world_rank_Shanghai          0
university_name              0
national_rank                0
overall_score_Shanghai    3794
alumni_score                 0
award_score                  0
hici_score                   0
ns_score                    20
publication_score            0
pcp_score                    0
year                         0
dtype: int64

In [23]:
#Locate the rows with missing values for ns_score

df_shanghai.loc[df_shanghai['ns_score'].isnull()]

world_rank_Shanghai                                   university_name  \
217              203-300  London School of Economics and Political Science   
348              301-400                     Stockholm School of Economics   
718              201-300  London School of Economics and Political Science   
847              301-400                     Stockholm School of Economics   
1158             151-202  London School of Economics and Political Science   
1347             305-402                     Stockholm School of Economics   
1726             201-302  London School of Economics and Political Science   
1963             402-503                     Stockholm School of Economics   
2226             201-302  London School of Economics and Political Science   
2466             402-501                     Stockholm School of Economics   
2730             201-300  London School of Economics and Political Science   
2859             301-400                     Stockholm School of Economics   
3122             102-150  London School of Economics and Political Science   
3364             301-400                     Stockholm School of Economics   
3464             401-500                                Tilburg University   
4006             101-150  London School of Economics and Political Science   
4340             401-500                     Stockholm School of Economics   
4347             401-500                                Tilburg University   
4505             101-150  London School of Economics and Political Science   
4843             401-500                     Stockholm School of Economics   

     national_rank  overall_score_Shanghai  alumni_score  award_score  \
217          20-30                     NaN          19.8          0.0   
348          10-11                     NaN           0.0         17.1   
718          23-33                     NaN          19.1          0.0   
847          10-11                     NaN           0.0         16.7   
1158         16-23                     NaN          18.6          0.0   
1347            10                     NaN           0.0         16.7   
1726         23-33                     NaN          17.7          0.0   
1963         10-11                     NaN           0.0         16.7   
2226         24-33                     NaN          17.3          0.0   
2466         10-11                     NaN           0.0         16.7   
2730         20-30                     NaN          16.9          0.0   
2859            10                     NaN           0.0         16.6   
3122         11-15                     NaN          22.8         16.2   
3364          9-10                     NaN           0.0         16.6   
3464            13                     NaN           0.0          0.0   
4006          9-17                     NaN          20.4         16.3   
4340            11                     NaN           0.0         16.3   
4347            13                     NaN           0.0          0.0   
4505         10-17                     NaN          19.9         16.3   
4843            11                     NaN           0.0         16.3   

      hici_score  ns_score  publication_score  pcp_score  year  
217         15.7       NaN               26.1       16.2  2005  
348          0.0       NaN               11.0       29.4  2005  
718         15.4       NaN               25.8       28.6  2006  
847          0.0       NaN               10.8       29.4  2006  
1158        16.6       NaN               25.3       28.4  2007  
1347         0.0       NaN               10.4       28.8  2007  
1726        16.3       NaN               26.4       17.2  2008  
1963         0.0       NaN               10.3       29.2  2008  
2226        14.5       NaN               26.1       28.5  2009  
2466         0.0       NaN               10.7       29.8  2009  
2730        16.1       NaN               26.0       24.8  2010  
2859         0.0       NaN               10.7       

There are 20 rows with missing values for the 'ns_score' column. However, the majority of the other columns for these rows have significant values so I will not be deleting these rows for now. 

In the end, the only columns with missing values will be 'overall_score_Shanghai' and 'ns_score'.

### Addressing Duplicate Values

In [29]:
df_dups_2 = df_shanghai[df_shanghai.duplicated()]

In [30]:
df_dups_2

Empty DataFrame
Columns: [world_rank_Shanghai, university_name, national_rank, overall_score_Shanghai, alumni_score, award_score, hici_score, ns_score, publication_score, pcp_score, year]
Index: []

There are no duplicate values in this dataframe.

### Summary Statistics

In [31]:
df_shanghai.describe()

overall_score_Shanghai  alumni_score  award_score   hici_score  \
count             1101.000000   4895.000000   4895.00000  4895.000000   
mean                36.383470      9.161757      7.69191    16.221491   
std                 13.557186     14.142081     15.49411    14.382710   
min                 23.500000      0.000000      0.00000     0.000000   
25%                 27.400000      0.000000      0.00000     7.300000   
50%                 31.300000      0.000000      0.00000    12.600000   
75%                 41.800000     15.600000     13.40000    21.700000   
max                100.000000    100.000000    100.00000   100.000000   

          ns_score  publication_score    pcp_score         year  
count  4875.000000        4895.000000  4895.000000  4895.000000  
mean     16.078503          38.254648    21.242329  2009.657406  
std      12.511529          13.050809     9.254351     3.197698  
min       0.000000           7.300000     8.300000  2005.000000  
25%       8.000000          28.900000    15.600000  2007.000000  
50%      12.800000          36.000000    19.000000  2009.000000  
75%      19.800000          45.300000    24.500000  2012.000000  
max     100.000000         100.000000   100.000000  2015.000000

The counts for the columns are not all the same because of the missing values. However, these statistics could provide some insights and also show that the missing values in the 'ns_score' column are considered NaN's. Another thing to note is that the 'overall_score_Shanghai' column data is most likely skewed, since not only is it missing a significant number of values compared to the other columns, but it's likely that scores that were deemed too low were not reported at all. 

### Exporting the cleaned file

In [32]:
df_shanghai.to_csv(os.path.join(path, '02 Data','Prepared Data', 'shanghai_cleaned_final.csv'))